### PR TITLE
Copter: Notify the frame type

### PR DESCRIPTION
**Pull request by muramura**
I am assembling a multicopter quad VTAIL.
I have multiple multicopter quads.
I would like to know what type of multicopter I have as well.

Screenshot from 2020-08-30 21-04-41

### CODE REVIEW

File: ArduCopter/system.cpp [hunk 1]
```diff
@@ -516,6 +516,41 @@ const char* Copter::get_frame_string()
...
+        return "BetaFlightX";
+    case 13:
+        return "DJIX";
+    case 14:
+        return "ClockwiseX";
+    case 15:
+        return "I";
+    case 18:
+        return "BetaFlightXReversed";
+    default:
```
Comment (IamPete1) [inline ArduCopter/system.cpp:549]: of you remove default this is more likely to be kept up to date

Comment (IamPete1) [reply ArduCopter/system.cpp:549]: you will have to cast to the enum for that to work

File: ArduCopter/GCS_Mavlink.cpp [hunk 2]
```diff
@@ -529,7 +529,7 @@ bool GCS_MAVLINK_Copter::params_ready() const
 void GCS_MAVLINK_Copter::send_banner()
 {
     GCS_MAVLINK::send_banner();
-    send_text(MAV_SEVERITY_INFO, "Frame: %s", copter.get_frame_string());
+    send_text(MAV_SEVERITY_INFO, "Frame: %s  Type: %s", copter.get_frame_string(), copter.get_type_string());
```
Comment (IamPete1) [inline ArduCopter/GCS_Mavlink.cpp:532]: I don't think you need `Type:`, just `Quad V-Tail` would be OK I think

**Review by khancyr — COMMENTED** [review]
I am a bit on edge on this...it doesn't bring much info and waste some flash with strings... Better have the GCS getting the frame and type correctly from parameters.
And it would probably need an enum for the frame_type to avoid maintenance issue

File: ArduCopter/system.cpp (hunk 1 above)
Comment (muramura) [reply ArduCopter/system.cpp:549]: The FRAME_TYPE value does not have a range.
Undefined values can be set by GCS.
I was able to verify this with APM PLANNER2.
I would modify it to show the set value if an undefined value is set.

Screenshot from 2020-08-31 07-13-41

Comment (peterbarker) [reply ArduCopter/system.cpp:549]: @muramura `enum class` required at the least.  Handling the case where the parameter is out-of-range can be done *outside* the switch statement, allowing you to have no `default:` clause (ensuring someone adding a new frame must add the string.  The existing equivalent code for the mav-type is not good.


What's the byte-cost of this new code?

The function name also needs to be more specific.

Comment (muramura) [reply ArduCopter/system.cpp:550]: I have defined the frame type in ENUM.

Comment (muramura) [reply ArduCopter/system.cpp:550]: The type of the frame type variable is AP_Int8.
I don't think you can specify the type of ENUM in the config parameter.
I can set it to a value other than the ENUM value.
I was able to set it in APM PLANNER2.
I think defult is valid.

Comment (peterbarker) [reply ArduCopter/system.cpp:550]: > The type of the frame_type variable is AP_Int8.

Yes.

> I don't think you can specify the type of ENUM in the config parameter.

You cast it as part of the switch statement.

> I can set it to a value other than the ENUM value.

Yes, you can.

> I think defult is valid.

No, it is not.  We use the switch statement to ensure the programmer takes care of adding the frame string.

If the value is out of range then you handle that after the body of the switch statement by returning your "DEFAULT" string.

Comment (muramura) [reply ArduCopter/system.cpp:550]: Displaying the frame type can reduce the time required for the following steps.
1. start up the GCS.
2 Select the configuration parameter item with the mouse.
3 Key in FRAME_TYPE.

Comment (IamPete1) [reply ArduCopter/system.cpp:550]: We already have a enum https://github.com/ArduPilot/ardupilot/blob/abee62abc453da8f558cbd76d34ce7b301a2e6ce/libraries/AP_Motors/AP_Motors_Class.h#L49

Comment (muramura) [reply ArduCopter/system.cpp:550]: @IamPete1 san. I'll change it to motor_frame_type.

**Comment by muramura** [comment]
I think of SERVO assignments when I have a frame type for log analysis.
I think this PR works for log analysis as well.

Screenshot from 2020-09-12 12-03-11

**Comment by peterbarker** [comment]
@muramura how much flash does this cost us?

As @khancyr has pointed out, this information is already available directly from the parameter value.

Countering @khancyr 's argument is the fact we see fit to emit the motors class string already, and its useless without the additional information you're adding here, pretty much.

I think we should probably take this to a devcall.

**Comment by IamPete1** [comment]
How about moving this to AP_Motors so we get the message on QuadPlane too?

File: ArduCopter/Parameters.cpp [hunk 3]
```diff
@@ -339,7 +339,7 @@ const AP_Param::Info Copter::var_info[] = {
     // @Param: FRAME_TYPE
     // @DisplayName: Frame Type (+, X, V, etc)
     // @Description: Controls motor mixing for multicopters.  Not used for Tri or Traditional Helicopters.
-    // @Values: 0:Plus, 1:X, 2:V, 3:H, 4:V-Tail, 5:A-Tail, 10:Y6B, 11:Y6F, 12:BetaFlightX, 13:DJIX, 14:ClockwiseX, 15: I, 18: BetaFlightXReversed
+    // @Values: 0:Plus, 1:X, 2:V, 3:H, 4:V-Tail, 5:A-Tail, 6:PLUSREV 10:Y6B, 11:Y6F, 12:BetaFlightX, 13:DJIX, 14:ClockwiseX, 15:I, 16:NYT_PLUS 17:NYT_X 18:BetaFlightXReversed
```
Comment (IamPete1) [inline ArduCopter/Parameters.cpp:342]: NYT_PLUS  and NYT_X  are not valid Copter frames, only work on tailsitter Quadplanes, other wise you will end up spinning with no yaw control.

Comment (peterbarker) [reply ArduCopter/Parameters.cpp:342]: Use `@Values{something}` to limit the list here, and add heli in

File: ArduCopter/system.cpp [hunk 4]
```diff
@@ -516,6 +516,46 @@ const char* Copter::get_frame_string()
...
+    case AP_Motors::MOTOR_FRAME_TYPE_I:
+        return "I";
+    case AP_Motors::MOTOR_FRAME_TYPE_NYT_PLUS:
+        return "NoYawTorquePlus";
+    case AP_Motors::MOTOR_FRAME_TYPE_NYT_X:
+        return "NoYawTorqueX";
+    case AP_Motors::MOTOR_FRAME_TYPE_BF_X_REV:
+        return "BetaFlightXReversed";
+    }
+    return "UNKNOWN";
```
Comment (tridge) [inline ArduCopter/system.cpp:556]: i think this will show UNKNOWN for heli?

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by rmackay9** [comment]
In general I like this idea.  I think it could save some time during support calls because we could avoid having to look up the FRAME_TYPE parameter to figure out what type of vehicle it is.

I like @IamPete1's idea of adding this to AP_Motors instead of having the lookup in the Copter code.

**Comment by peterbarker** [comment]
@kd0aij is implementing all of this as suggested on this PR in his draft PR over here: https://github.com/ArduPilot/ardupilot/pull/15748/files

@kd0aij there are some comments in here about the way things might be formatted.

@muramura probably not much point duplicating the effort to put the type strings in.  Thanks for kicking this off - and we should definitely periodically poke @kd0aij so he continues on that PR :-)

**Pull request closed without merging**